### PR TITLE
planner: support cost model for Chunk encode format.

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve? 

Add a new cost model for Chunk format.

### What is changed and how it works?

1. Implement new AvgRowSize for Chunk format.
2. Use sessionVar.EnableChunkRPC to choose default cost or chunk cost. 

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

Release note

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -140,6 +140,34 @@ func (c *Column) AvgColSize(count int64, isKey bool) float64 {
...
+	case mysql.TypeTiny, mysql.TypeDouble, mysql.TypeDuration, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear, mysql.TypeEnum, mysql.TypeBit, mysql.TypeSet:
+		return 8
+	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+		return 20
+	case mysql.TypeNewDecimal:
+		return 40
+	}
+	// Keep two decimal place.
+	// Add 8 bytes for unfixed-len type's offsets.
+	return math.Round(float64(c.TotColSize)/float64(count)*100)/100 + 8
```
Comment (XuHuaiyu) [inline statistics/histogram.go]: What's  this for?

File: planner/core/task.go [hunk 2]
```diff
@@ -127,7 +127,11 @@ func (t *copTask) finishIndexPlan() {
 	t.indexPlanFinished = true
 	sessVars := t.indexPlan.SCtx().GetSessionVars()
 	// Network cost of transferring rows of index scan to TiDB.
-	t.cst += cnt * sessVars.NetworkFactor * t.tblColHists.GetAvgRowSize(t.indexPlan.Schema().Columns, true)
+	if sessVars.EnableChunkRPC {
```
Comment (XuHuaiyu) [inline planner/core/task.go]: This is not enough

We need to choose a path from
```
IndexScan_TypeDefault
TableScan_TypeDefault
IndexLookup_TypeDefault
IndexScan_TypeChunk
TableScan_TypeChunk
IndexLookup_TypeChunk
```

File: bindinfo/bind_test.go [hunk 3]
```diff
@@ -521,6 +521,6 @@ func (s *testSuite) TestAddEvolveTasks(c *C) {
 	domain.GetDomain(tk.Se).BindHandle().SaveEvolveTasksToStore()
 	rows := tk.MustQuery("show global bindings").Rows()
 	c.Assert(len(rows), Equals, 2)
-	c.Assert(rows[1][1], Equals, "SELECT /*+ USE_INDEX(@`sel_1` `test`.`t` )*/ * FROM `test`.`t` WHERE `a`>=4 AND `b`>=1 AND `c`=0")
+	c.Assert(rows[1][1], Equals, "SELECT /*+ USE_INDEX(@`sel_1` `test`.`t` `idx_c`)*/ * FROM `test`.`t` WHERE `a`>=4 AND `b`>=1 AND `c`=0")
```
Comment (eurekaka) [inline bindinfo/bind_test.go]: Update the comment in line 518 also.

File: cmd/explaintest/r/select.result [hunk 4]
```diff
@@ -267,30 +267,30 @@ insert t values(0,0,0);
 explain select distinct b from t group by a;
 id	count	task	operator info
 HashAgg_7	8000.00	root	group by:Column#2, funcs:firstrow(Column#2)
-└─StreamAgg_22	8000.00	root	group by:Column#12, funcs:firstrow(Column#11)
-  └─IndexReader_23	8000.00	root	index:StreamAgg_11
-    └─StreamAgg_11	8000.00	cop[tikv]	group by:Column#1, funcs:firstrow(Column#2)
-      └─IndexScan_21	10000.00	cop[tikv]	table:t, index:a, b, c, range:[NULL,+inf], keep order:true, stats:pseudo
+└─HashAgg_15	8000.00	root	group by:Column#8, funcs:firstrow(Column#7)
```
Comment (eurekaka) [inline cmd/explaintest/r/select.result]: Is this change reasonable?

File: planner/core/find_best_task.go [hunk 5]
```diff
@@ -1091,12 +1107,22 @@ func (ds *DataSource) getOriginalPhysicalTableScan(prop *property.PhysicalProper
...
 	var rowSize float64
-	if ts.StoreType == kv.TiKV {
-		rowSize = ds.TblColHists.GetTableAvgRowSize(ds.TblCols, ts.StoreType, true)
+	if ds.ctx.GetSessionVars().EnableChunkRPC {
+		if ts.StoreType == kv.TiKV {
+			rowSize = ds.TblColHists.GetTableAvgRowSizeChunkFormat(ds.TblCols, ts.StoreType, true)
+		} else {
+			// If `ds.handleCol` is nil, then the schema of tableScan doesn't have handle column.
+			// This logic can be ensured in column pruning.
+			rowSize = ds.TblColHists.GetTableAvgRowSizeChunkFormat(ts.Schema().Columns, ts.StoreType, ds.handleCol != nil)
```
Comment (eurekaka) [inline planner/core/find_best_task.go]: Has TiFlash supported Arrow format returning now?

File: statistics/histogram.go [hunk 6]
```diff
@@ -140,6 +140,34 @@ func (c *Column) AvgColSize(count int64, isKey bool) float64 {
 	return math.Round(float64(c.TotColSize)/float64(count)*100) / 100
 }
 
+// AvgColSizeChunkFormat is the average column size of the histogram. These sizes are derived from function `Encode`
+// and `DecodeToChunk`, so we need to update them if those 2 functions are changed.
```
Comment (eurekaka) [inline statistics/histogram.go]: This comment is not suitable here?

File: statistics/histogram.go [hunk 7]
```diff
@@ -140,6 +140,34 @@ func (c *Column) AvgColSize(count int64, isKey bool) float64 {
...
+func (c *Column) AvgColSizeChunkFormat(count int64) float64 {
+	if count == 0 {
+		return 0
+	}
+	// Note that, if the handle column is encoded as value, instead of key, i.e,
+	// when the handle column is in a unique index, the real column size may be
+	// smaller than 8 because it is encoded using `EncodeVarint`. Since we don't
+	// know the exact value size now, use 8 as approximation.
+	if c.IsHandle {
```
Comment (eurekaka) [inline statistics/histogram.go]: This is not applicable either? we can remove this check here actually?

File: statistics/histogram.go [hunk 8]
```diff
@@ -140,6 +140,34 @@ func (c *Column) AvgColSize(count int64, isKey bool) float64 {
...
+	}
+	switch c.Histogram.Tp.Tp {
+	case mysql.TypeFloat:
+		return 4
+	case mysql.TypeTiny, mysql.TypeDouble, mysql.TypeDuration, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong, mysql.TypeYear, mysql.TypeEnum, mysql.TypeBit, mysql.TypeSet:
+		return 8
+	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+		return 20
+	case mysql.TypeNewDecimal:
+		return 40
```
Comment (eurekaka) [inline statistics/histogram.go]: Use variable like `MyDecimalStructSize` please.

File: planner/core/exhaust_physical_plans.go [hunk 9]
```diff
@@ -684,7 +684,12 @@ func (p *LogicalJoin) constructInnerTableScanTask(
 	for i := range ds.stats.Cardinality {
 		ds.stats.Cardinality[i] = 1
 	}
-	rowSize := ds.TblColHists.GetTableAvgRowSize(ds.TblCols, ts.StoreType, true)
+	var rowSize float64
+	if p.ctx.GetSessionVars().EnableChunkRPC {
+		rowSize = ds.TblColHists.GetTableAvgRowSizeChunkFormat(ds.TblCols, ts.StoreType, true)
+	} else {
+		rowSize = ds.TblColHists.GetTableAvgRowSize(ds.TblCols, ts.StoreType, true)
```
Comment (eurekaka) [inline planner/core/exhaust_physical_plans.go]: Can we check `EnableChunkRPC` in functions like `GetTableAvgRowSize`, then we don't need to add this check everywhere?

File: statistics/histogram.go (hunk 8 above)
Comment (SunRunAway) [reply statistics/histogram.go]: There're too many magic number here, could we use export and use `chunk.getFixedLen` here?

Comment (wshwsh12) [reply statistics/histogram.go]: Addressed.

File: planner/core/exhaust_physical_plans.go (hunk 9 above)
Comment (wshwsh12) [reply planner/core/exhaust_physical_plans.go]: Addressed.

File: statistics/histogram.go (hunk 7 above)
Comment (wshwsh12) [reply statistics/histogram.go]: Addressed. I remove it.

File: planner/core/find_best_task.go (hunk 5 above)
Comment (wshwsh12) [reply planner/core/find_best_task.go]: Yes.

File: statistics/histogram.go (hunk 1 above)
Comment (wshwsh12) [reply statistics/histogram.go]: I copy this line from `AvgColSize` . This line calculate the avg column size for unfixed-len (I think) and I add 8 bytes for unfixed-len type's offsets.

File: planner/core/task.go (hunk 2 above)
Comment (wshwsh12) [reply planner/core/task.go]: I don't think so. If the plan is decided, then the EncodeType is stable.  
For example, If we want choose a encodeType for a TableScan, calculate cost(TableScan_TypeDefault) and cost(TableScan_TypeChunk), select the smaller one and set the option.

File: bindinfo/bind_test.go (hunk 3 above)
Comment (wshwsh12) [reply bindinfo/bind_test.go]: I think this change is resonable.  How to change the comment?

File: statistics/histogram.go (hunk 6 above)
Comment (wshwsh12) [reply statistics/histogram.go]: Why?

File: cmd/explaintest/r/select.result (hunk 4 above)
Comment (wshwsh12) [reply cmd/explaintest/r/select.result]: No. I fix it.

**Comment by eurekaka** [comment]
@SunRunAway PTAL

**Comment by XuHuaiyu** [comment]
The CPU cost of decoding is still not considered?

**Comment by wshwsh12** [comment]
> The CPU cost of decoding is still not considered?

File: statistics/histogram.go (hunk 1 above)
Comment (SunRunAway) [reply statistics/histogram.go]: `TotColSize` includes the size of `LEN` and `BYTE`, so I think we need to minus the size of `LEN` here?
And I've notice it minus 1 but not the actual size of `LEN` in `AvgColSizeListInDisk`, could you help fix it either?

File: statistics/table.go [hunk 10]
```diff
@@ -688,9 +690,17 @@ func (coll *HistColl) GetAvgRowSize(cols []*expression.Column, isEncodedKey bool
 			}
 			// We differentiate if the column is encoded as key or value, because the resulted size
 			// is different.
-			size += colHist.AvgColSize(coll.Count, isEncodedKey)
+			if sessionVars.EnableChunkRPC {
+				size += colHist.AvgColSizeChunkFormat(coll.Count)
```
Comment (SunRunAway) [inline statistics/table.go]: If a cost is describing scan factor in TiKV, should we always use `colHist.AvgColSize` but not `colHist.AvgColSizeChunkFormat`?
Like https://github.com/pingcap/tidb/pull/13461/files#diff-152a0fbc96916798e8bccdb90988517fR689

File: statistics/histogram.go (hunk 1 above)
Comment (wshwsh12) [reply statistics/histogram.go]: `TotColSize` doesn't includes the size of `LEN`.

File: statistics/table.go (hunk 10 above)
Comment (wshwsh12) [reply statistics/table.go]: Addressed. PTAL again. @SunRunAway

File: statistics/histogram.go (hunk 1 above)
Comment (SunRunAway) [reply statistics/histogram.go]: `TotColSize` does not have a decent comment to demostrate this, but I found a comment.
https://github.com/pingcap/tidb/blob/5a589c9994b568368e185602e54bbcfffb0f4c44/statistics/handle/handle_test.go#L215

**Pull request merged**